File: 19.py
#!/bin/python3

from core.skel import *

sample = """
--- scanner 0 ---
404,-588,-901
528,-643,409
-838,591,734
390,-675,-793
-537,-823,-458
-485,-357,347
-345,-311,381
-661,-816,-575
-876,649,763
-618,-824,-621
553,345,-567
474,580,667
-447,-329,318
-584,868,-557
544,-627,-890
564,392,-477
455,729,728
-892,524,684
-689,845,-530
423,-701,434
7,-33,-71
630,319,-379
443,580,662
-789,900,-551
459,-707,401

--- scanner 1 ---
686,422,578
605,423,415
515,917,-361
-336,658,858
95,138,22
-476,619,847
-340,-569,-846
567,-361,727
-460,603,-452
669,-402,600
729,430,532
-500,-761,534
-322,571,750
-466,-666,-811
-429,-592,574
-355,545,-477
703,-491,-529
-328,-685,520
413,935,-424
-391,539,-444
586,-435,557
-364,-763,-893
807,-499,-711
755,-354,-619
553,889,-390

--- scanner 2 ---
649,640,665
682,-795,504
-784,533,-524
-644,584,-595
-588,-843,648
-30,6,44
-674,560,763
500,723,-460
609,671,-379
-555,-800,653
-675,-892,-343
697,-426,-610
578,704,681
493,664,-388
-671,-858,530
-667,343,800
571,-461,-707
-138,-166,112
-889,563,-600
646,-828,498
640,759,510
-630,509,768
-681,-892,-333
673,-379,-804
-742,-814,-386
577,-820,562

--- scanner 3 ---
-589,542,597
605,-692,669
-500,565,-823
-660,373,557
-458,-679,-417
-488,449,543
-626,468,-788
338,-750,-386
528,-832,-391
562,-778,733
-938,-730,414
543,643,-506
-524,371,-870
407,773,750
-104,29,83
378,-903,-323
-778,-728,485
426,699,580
-438,-605,-362
-469,-447,-387
509,732,623
647,635,-688
-868,-804,481
614,-800,639
595,780,-596

--- scanner 4 ---
727,592,562
-293,-554,779
441,611,-461
-714,465,-776
-743,427,-804
-660,-479,-426
832,-632,460
927,-485,-438
408,393,-506
466,436,-512
110,16,151
-258,-428,682
-393,719,612
-211,-452,876
808,-476,-593
-575,615,604
-485,667,467
-680,325,-822
-627,-443,-432
872,-547,-609
833,512,582
807,604,487
839,-516,451
891,-625,532
-652,-548,-490
30,-46,-14
"""
answer1 = None
answer2 = None

def parse(input):
    lines = [parse_fields(x,digits+alpha+'-') for x in input.splitlines() if x]
    scanner = []
    for x in lines:
        if x[1] == "scanner":
            scanner.append(set())
        else:
            scanner[-1].add(tuple([int(a) for a in x]))

    return scanner

# Rotate given scanner around x axis (y becomes z)
def rotate_x(s):
    return set([ (x,z,-y) for x,y,z in s])

# Rotate given scanner around x axis (y becomes z)
def rotate_y(s):
    return set([ (-z,y,x) for x,y,z in s])

# Rotate given scanner around x axis (y becomes z)
def rotate_z(s):
    return set([ (-y,x,z) for x,y,z in s])

def translate(s, pos):
    X,Y,Z = pos
    return set([ (x+X, y+Y, z+Z) for x,y,z in s] )

sloc = []
# Returns set of beacons from t rotated/translated into s coord space, if overlap detected
def overlap(s,t):

    for sb in s:
        sx, sy, sz = sb
        for tb in t:
            tx, ty, tz = tb
            # assume beacons sb and tb are the same beacon in the same scanner orientation
            # Find location of scanner t relative to s
            # Determine if there are 12 matching beacons in this location
            torigin = (sx-tx, sy-ty, sz-tz)
            ts = translate(t, torigin)

            common = ts.intersection(s)
            if len(common) >= 12:
                print("Found map; scanner is at ", torigin)
                sloc.append(torigin)
                return ts
    return set()

def rotated(s, memo={}):
    already = set()
    r = frozenset(s)
    if r in memo:
        for x in memo[r]:
            yield x
    else:
        # Try every rotation
        for rx in range(4):
            r = rotate_x(r)
            for ry in range(4):
                r = rotate_y(r)
                for rz in range(4):
                    r = frozenset(rotate_z(r))
                    if r in already:
                        continue
                    already.add(r)

                    yield r
        memo[r] = already

def rotate(s,t):
    memo = set()
    for r in rotated(t):
        common = overlap(s,r)
        if common:
            return common
    return set()

def manhattan(a,b):
    x,y,z = a
    X,Y,Z = b

    return abs(x-X) + abs(y-Y) + abs(z-Z)

def part1(scanner, part2=False):
    global sloc
    map = scanner[0]
    sloc = [(0,0,0)]

    # Beacons rotated into place relative to scanner 0
    fixed = scanner[0:1]
    unknown = scanner[1:]

    while True:
        found = []
        new = []
        for x in range(len(unknown)):
            t = frozenset(unknown[x])
            for s in fixed:
                match = rotate(s, t)
                if match:
                    found.append(x)
                    new.append(match)
                    map |= match
                    print(f"Matched scanner {x+1} of {len(unknown)}  beacons={len(map)}  scanners={len(sloc)}")
                    break

        found.reverse()
        print(found)
        for x in found:
            del unknown[x]
        fixed = new
        if not found:
            break

    c = 0
    for s in sloc:
        c+=1
        print(f"{c}: {s}")
    dist = []
    for i in range(len(sloc)-1):
        for j in range(i+1,len(sloc)):
            dist.append((manhattan(sloc[i], sloc[j]), sloc[i], sloc[j]))
    print("  ", max(dist))

    return len(map)


def part2(input):
    pass


from aocd import data
runAll(sample, data, parse, part1, part2, answer1, answer2)

File: 19.cpp
#include <iostream>
#include <fstream>
#include <sstream>

#include <vector>
#include <set>
#include <unordered_set>
#include <algorithm>

using namespace std;

struct Point {
    int x, y, z;

    bool operator==(const Point& rhs) const {
        return rhs.x == x && rhs.y ==y && rhs.z == z;
    }

    bool operator<(const Point& rhs) const {
        return rhs.x < x ||
            (rhs.x == x && rhs.y < y) ||
            (rhs.y == y && rhs.z < z);
    }

    Point operator+(const Point& rhs) const {
        return {rhs.x + x, rhs.y + y, rhs.z + z};
    }

    Point operator-(const Point& rhs) const {
        return {x - rhs.x , y - rhs.y , z - rhs.z };
    }

    int manhattan_dist(Point q) {
        auto diff = *this - q;
        return abs(diff.x) + abs(diff.y) + abs(diff.z);
    }
};

ostream& operator<<(ostream& os, const Point& p)
{
    os << p.x << ',' << p.y << ',' << p.z;
    return os;
}


using Beacons = set<Point>;
struct Scanner {
    int id;
    Point offset;
    Beacons beacons;
    set<Beacons> rotations;
    set<int> fingerprint;

    Scanner() {
        rotations.insert({});
    }

    // Fingerprint is a quick and dirty discriminator for potential overlapping scanner cubes.  An overlapping cube has
    // to have at least 12 points in common.  This fingerprint holds all the distances between every pair of beacons. If
    // our cube overlaps another, we should have at least 11 + 10 + 9 + ... + 2 + 1 = 12 * 5.5 = 66 distances in common.
    // But we keep a set and some distances may be duplicated. So this could be quite wrong. Fortunately for my puzzle
    // input, it was not.
    // Someone else had the clever idea to record triangles (3 distances?).  This would probably be safer.
    void gen_fingerprint() {
        for (auto a : beacons) {
            for (auto b : beacons) {
                auto d = a.manhattan_dist(b);
                if (d) fingerprint.insert(d);
            }
        }
    }

    bool maybe_has_overlap(Scanner const &rhs) const {
        set<int> result;
        set_intersection(fingerprint.begin(), fingerprint.end(), rhs.fingerprint.begin(), rhs.fingerprint.end(),
                        std::inserter(result, result.begin()));
        return result.size() > 50;
    }
};

using Scanners = vector<Scanner>;

Scanners parse() {
    Scanners inp;
    ifstream infile("19.txt");
    // ifstream infile("sample.txt");
    string line;

    while (getline(infile, line)) {
        if (line.empty()) {
            continue;
        } else if (line.find("scanner") != string::npos) {
            inp.push_back({});
            continue;
        }

        stringstream ss;
        ss << line;
        int x, y, z;
        char comma;

        ss >> x >> comma >> y >> comma >> z;
        // cout << line << "   " << x << ", " << y << ", " <<z  <<"\n";
        auto beac = inp.back().beacons.insert({x, y, z});
        inp.back().id = inp.size()-1;
    }
    return inp;
}

Point rotate_x(Point p) {
    return {p.x, p.z, -p.y};
}
Point rotate_y(Point p) {
    return {-p.z, p.y, p.x};
}
Point rotate_z(Point p) {
    return {-p.y, p.x, p.z};
}
Beacons rotate_x(Beacons const & b) {
    Beacons nb;
    for ( auto const p : b )
        nb.insert(rotate_x(p));
    return nb;
}
Beacons rotate_y(Beacons const & b) {
    Beacons nb;
    for ( auto const p : b )
        nb.insert(rotate_y(p));
    return nb;
}
Beacons rotate_z(Beacons const & b) {
    Beacons nb;
    for ( auto const p : b )
        nb.insert(rotate_z(p));
    return nb;
}

// Generate all the rotations of the beacons in the scanner range
void rotate_all(Scanner & scan) {
    Beacons const & b = scan.beacons;

    Beacons r = b;
    // cout << "rotate_all " << r.size() << endl;
    for (int x=0; x < 4; x++) {
        for (int y=0; y < 4; y++) {
            for (int z=0; z < 4; z++) {
                scan.rotations.insert(r);
                r = rotate_z(r);
            }
            r = rotate_y(r);
        }
        r = rotate_x(r);
    }
}

// This is complex (big-O) because we repeatedly translate beacons for trials.  A faster
// method is to fingerprint the beaconspace by calculating distances between beacons and then
// deciding if there are 12 beacons' worth of common distances.
Beacons translate(Beacons & b, Point tx) {
    Beacons ans;
    // cout << tx << endl;
    for (auto const p : b) {
        // cout << "  " << p << "   ->   " << p + tx << endl;
        ans.insert(p + tx);
    }
    return ans;
}

// If two scanners have 12 beacons in common when translated and rotated, return the whole translated and rotated set from t
Scanner overlap(Scanner const & s, Scanner const & t) {
    if (!s.maybe_has_overlap(t))
        return {};

    auto const & sb = s.beacons;
    for (auto sp : sb) {
        for (auto tb: t.rotations) {
            for (auto tp : tb) {
                auto torigin = sp - tp;
                auto trial = translate(tb, torigin);
                Beacons result;
                set_intersection(sb.begin(), sb.end(), trial.begin(), trial.end(),
                                std::inserter(result, result.begin()));
                if (result.size() >= 12) {
                    cout << "Found match between "<< s.id << "-"<< t.id << " at " << torigin << "\n";
                    Scanner ns = t;
                    ns.beacons = std::move(trial);
                    ns.offset = torigin;
                    ns.rotations.clear();
                    return ns;
                }
            }
        }
    }
    // cout << "  no match: "<< s.id << "-"<< t.id << "\n";
    return {};
}

int main() {
    auto scanners = parse();
    for (auto & s: scanners) {
        rotate_all(s);
        s.gen_fingerprint();
    }

    // Pick a scanner to be the "real" coordinates
    auto const & s0 = scanners[0];

    vector<Scanner> fixed = {s0};
    vector<Scanner> unknown;
    for (int i = 1; i < scanners.size(); i++) {
        unknown.push_back(scanners[i]);
    }

    Beacons map = s0.beacons;

    // Hack: store scanner locations in this faux scanner structure so we can use fingerprint for part 2.
    Scanner scanner_positions;
    while (unknown.size()) {
        vector<Scanner> new_unknown;
        vector<Scanner> new_fixed;
        for (auto const & t : unknown) {
            bool found = false;
            for (auto const & s : fixed) {

                auto match = overlap(s, t);

                if (!match.beacons.empty()) {
                    map.insert(match.beacons.begin(), match.beacons.end());
                    scanner_positions.beacons.insert(match.offset);
                    new_fixed.push_back(std::move(match));
                    found = true;
                    break;
                }
            }
            if (!found) new_unknown.push_back(std::move(t));
        }
        unknown = std::move(new_unknown);
        fixed = std::move(new_fixed);
        if (fixed.empty() && !unknown.empty()) {
            cout << "Error: still have " << unknown.size() << " scanners to map\n";
        }
    }

    scanner_positions.gen_fingerprint();
    cout << map.size() << " " << *scanner_positions.fingerprint.crbegin() << endl;
    return 0;
}